tokio/src/runtime/basic_scheduler.rs

```diff
@@ -117,13 +133,65 @@ where
         self.spawner.spawn(future)
     }
 
-    pub(crate) fn block_on<F>(&mut self, future: F) -> F::Output
-    where
-        F: Future,
-    {
+    pub(crate) fn block_on<F: Future>(&self, future: F) -> F::Output {
+        // If we can steal the dedicated parker then lets block_on that
+        // Otherwise, we'll block_on the future and attempt to steal the
+        // parker later, if we can.
+        if let Some(mut inner) = self.take_inner() {
+            inner.block_on(future)
+        } else {
+            let enter = crate::runtime::enter(false);
+
+            let mut park = ParkThread::new();
+            let waker = park.unpark().into_waker();
+            let mut cx = std::task::Context::from_waker(&waker);
+
+            pin!(future);
+
+            loop {
+                if let Ready(v) = crate::coop::budget(|| future.as_mut().poll(&mut cx)) {
+                    return v;
+                }
+
+                // Check if we can steal the dedicated parker P.
+                //
+                // TODO: Consider using an atomic load here intead of locking
+                // the mutex.
+                if let Some(mut inner) = self.take_inner() {
+                    // We will enter again on in the inner implementation below
+                    drop(enter);
+                    return inner.block_on(future);
+                } else {
+                    let waker = park.unpark().into_waker();
+                    let mut lock = self.steal_queue.lock().unwrap();
```

Comment:
This looks like there would be a race condition between trying to take "inner" and pushing the waker in the VecDeque. `Inner` could be returned before the waker is pushed into the queue.

This should be resolved by using a single mutex.